**Calculations for the given CNN Model**

In [5]:
input_height = 128
input_width = 128
input_channels = 1  

conv1_filters = 16
conv1_kernel_size = 5
conv1_stride = 1
conv1_padding = 2  

conv1_output_height = (input_height + 2 * conv1_padding - conv1_kernel_size) // conv1_stride + 1
conv1_output_width = (input_width + 2 * conv1_padding - conv1_kernel_size) // conv1_stride + 1
conv1_output_channels = conv1_filters

maxpool1_size = 2
maxpool1_stride = 2

maxpool1_output_height = (conv1_output_height - maxpool1_size) // maxpool1_stride + 1
maxpool1_output_width = (conv1_output_width - maxpool1_size) // maxpool1_stride + 1
maxpool1_output_channels = conv1_output_channels

conv2_filters = 32
conv2_kernel_size = 3
conv2_stride = 1
conv2_padding = 1  

conv2_output_height = (maxpool1_output_height + 2 * conv2_padding - conv2_kernel_size) // conv2_stride + 1
conv2_output_width = (maxpool1_output_width + 2 * conv2_padding - conv2_kernel_size) // conv2_stride + 1
conv2_output_channels = conv2_filters

maxpool2_size = 2
maxpool2_stride = 2

maxpool2_output_height = (conv2_output_height - maxpool2_size) // maxpool2_stride + 1
maxpool2_output_width = (conv2_output_width - maxpool2_size) // maxpool2_stride + 1
maxpool2_output_channels = conv2_output_channels

fc1_neurons = 256

output_classes = 10

params_conv1 = conv1_filters * (conv1_kernel_size**2) * input_channels + conv1_filters
params_conv2 = conv2_filters * (conv2_kernel_size**2) * conv1_output_channels + conv2_filters
params_fc1 = maxpool2_output_height * maxpool2_output_width * maxpool2_output_channels * fc1_neurons + fc1_neurons
params_output = fc1_neurons * output_classes + output_classes

print("Q.1. Output volume dimensions after each layer:")
print(f"Conv1 output: ({conv1_output_height}, {conv1_output_width}, {conv1_output_channels})")
print(f"MaxPool1 output: ({maxpool1_output_height}, {maxpool1_output_width}, {maxpool1_output_channels})")
print(f"Conv2 output: ({conv2_output_height}, {conv2_output_width}, {conv2_output_channels})")
print(f"MaxPool2 output: ({maxpool2_output_height}, {maxpool2_output_width}, {maxpool2_output_channels})")
print()
print("Q.2.Total number of parameters in the CNN (excluding biases):")
print(f"Conv1 params: {params_conv1}")
print(f"Conv2 params: {params_conv2}")
print(f"FC1 params: {params_fc1}")
print(f"Output params: {params_output}")


Q.1. Output volume dimensions after each layer:
Conv1 output: (128, 128, 16)
MaxPool1 output: (64, 64, 16)
Conv2 output: (64, 64, 32)
MaxPool2 output: (32, 32, 32)

Q.2.Total number of parameters in the CNN (excluding biases):
Conv1 params: 416
Conv2 params: 4640
FC1 params: 8388864
Output params: 2570


(Rough Calculation Results):

Q1. Output volume dimensions after each layer:
Conv1 output: (128, 128, 16)
MaxPool1 output: (64, 64, 16)
Conv2 output: (64, 64, 32)
MaxPool2 output: (32, 32, 32)

Q2. Total number of parameters in the CNN (excluding biases):
Conv1 params: 416
Conv2 params: 4640
FC1 params: 8388864
Output params: 2570